### ISSUE TITLE
Reaper cache removal needs lock?

### ISSUE THREAD

**Issue by gbiddison** [jstasiak/python-zeroconf#38]
```python
Exception in thread Thread-29:
Traceback (most recent call last):
  File "/usr/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.4/dist-packages/zeroconf.py", line 919, in run
    for record in self.zc.cache.entries():
  File "/usr/local/lib/python3.4/dist-packages/zeroconf.py", line 793, in entries
    return reduce(lambda a, b: a + b, self.cache.values())
RuntimeError: dictionary changed size during iteration
```

**Comment by gbiddison** [comment]
Occasionally seeing this error during steady state operation with new zerconf services appearing and/or disappearing.

**Comment by gbiddison** [comment]
I'm guessing this kills the Reaper thread based on the lack of try / except; funny thing is this doesn't cause any obvious problem with the system.

**Comment by gbiddison** [comment]
working on PR for proposed patch:

diff --git a/zeroconf.py b/zeroconf.py
index 627e571..9a9d02d 100644
--- a/zeroconf.py
+++ b/zeroconf.py
@@ -916,7 +916,8 @@ class Reaper(threading.Thread):
             if _GLOBAL_DONE:
                 return
             now = current_time_millis()
-            for record in self.zc.cache.entries():
+            entries = list(self.zc.cache.entries())
+            for record in entries:
                 if record.is_expired(now):
                     self.zc.update_record(now, record)
                     self.zc.cache.remove(record)

Unfortunately I can't make a test case for this, since it's intermittent.  I will observe it for a while to see if this solves the problem.

**Comment by gbiddison** [comment]
I should add that this exception has been observed in python 3.4 on both OSX and Ubuntu 14.04.3

**Comment by gbiddison** [comment]
final working solution is a patch on the inner loop:
```diff
index 1018fe0..5ff321a 100644
--- a/zeroconf.py
+++ b/zeroconf.py
@@ -790,7 +790,8 @@ def entries(self):
         if not self.cache:
             return []
         else:
-            return reduce(lambda a, b: a + b, self.cache.values())
+            values = list(self.cache.values())
+            return reduce(lambda a, b: a + b, values)
```

**Comment by gbiddison** [comment]
Also see this stack overflow about refactoring this to [item for sublist in l for item in sublist] since this just appears to be a list flattening operation.

http://stackoverflow.com/questions/952914/making-a-flat-list-out-of-list-of-lists-in-python

**Comment by jstasiak** [comment]
Fix merged as fd9d531f294e7fa5b9b934f192b061f56eaf1d37, thank you for the report and the patch.
